### PR TITLE
Otlp exporter

### PR DESCRIPTION
**Pull request by 15skumar**
This PR adds an implementation of OpenTelemetry Exporting of metrics through the OpenTelemetry protocol (OTLP). Changes are: a new OtlpMetricsExporter struct which handles exporting metrics to an OTLP endpoint, and integration of the OTLP exporter with the existing metrics system. 

Testing:
I tested the implementation with a test otlp_metrics() in metrics.rs and ran a docker container running the OpenTelemetry Collector at the default port

docker run -d --name otel-collector \
  -p 4318:4318 -p 4317:4317 \
  -v $(pwd)/collector-config.yaml:/etc/otelcol/config.yaml \
  otel/opentelemetry-collector-contrib:latest

Once I ran the test, I verified that the test metrics can be viewed in the collector logs. (viewed using 'docker logs otel-collector'). Here is a screenshot of an example of a test metric collected at the endpoint: 
Screenshot 2025-06-18 at 15 32 16

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: collector-config.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
+receivers:
```
Comment (muddyfish) [inline collector-config.yaml]: Should this be committed?

File: mountpoint-s3-fs/src/metrics.rs [hunk 2]
```diff
@@ -90,12 +93,27 @@ fn poll_process_metrics(sys: &mut System) {
 #[derive(Debug)]
 struct MetricsSink {
     metrics: DashMap<Key, Metric>,
+    otlp_exporter: Option<OtlpMetricsExporter>, // optional OTLP exporter
 }
 
 impl MetricsSink {
-    fn new() -> Self {
+    fn new(otlp_endpoint: Option<&str>) -> Self {
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: Perhaps we should add a config struct for this so it's easier to add/change fields

File: mountpoint-s3-fs/src/metrics.rs [hunk 3]
```diff
@@ -90,12 +93,27 @@ fn poll_process_metrics(sys: &mut System) {
...
+        // Initialise the OTLP exporter if an endpoint is provided
+        // check syntax and if error case works
+        let otlp_exporter = otlp_endpoint.and_then(|endpoint| match OtlpMetricsExporter::new(endpoint) {
+            Ok(exporter) => {
+                tracing::info!("OpenTelemetry metrics export enabled to {}", endpoint);
+                Some(exporter)
+            }
+            Err(e) => {
+                tracing::error!("Failed to initialise OTLP exporter: {}", e);
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: Do we want to continue or error out in this case?

File: mountpoint-s3-fs/src/metrics.rs [hunk 4]
```diff
@@ -310,4 +359,85 @@ mod tests {
             }
         });
     }
+
+    #[test]
+    #[ignore] // Mark as ignored so it doesn't run in normal test runs
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: This test appears to just export metrics and then not have any validation steps. Ideally tests should fail if they don't work properly

File: mountpoint-s3-fs/src/metrics.rs [hunk 5]
```diff
@@ -310,4 +359,85 @@ mod tests {
...
+            .with_level(true)
+            .with_file(true)
+            .with_line_number(true)
+            .with_test_writer()
+            .set_default();
+
+        info!("Starting OTLP metrics test...");
+
+        // Initialize metrics with an OTLP endpoint
+        let sink = Arc::new(MetricsSink::new(Some("http://localhost:4318/v1/metrics")));
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: Nit: This should probably be an environment variable

File: mountpoint-s3-fs/src/metrics.rs [hunk 6]
```diff
@@ -310,4 +359,85 @@ mod tests {
...
+            histogram.record(10.0);
+            histogram.record(20.0);
+            histogram.record(30.0);
+            info!("Recorded histogram values: 10.0, 20.0, 30.0");
+
+            // Publish metrics immediately to verify initial values
+            info!("Publishing initial metrics...");
+            sink.publish();
+
+            // Sleep to allow metrics to be exported
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs:471]: Exported from where?

File: mountpoint-s3-fs/src/metrics_otel.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            .with_protocol(Protocol::HttpBinary)
+            .with_endpoint(endpoint)
+            .build()?;
+
+        // Create a meter provider with the OTLP Metric Exporter that will collect and export metrics at regular intervals
+        let meter_provider = opentelemetry_sdk::metrics::SdkMeterProvider::builder()
+            // The default interval is 60 seconds so we use a PeriodicReader to allow us to specify a custom interval duration
+            .with_reader(
+                opentelemetry_sdk::metrics::PeriodicReader::builder(exporter)
+                    .with_interval(Duration::from_secs(5)) // This sets a 5-second export interval
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics_otel.rs]: Should be configurable

File: mountpoint-s3-fs/src/metrics_otel.rs [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                    .build(),
+            )
+            .build();
+        global::set_meter_provider(meter_provider);
+
+        // Get a meter
+        let meter = global::meter("mountpoint-s3");
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics_otel.rs:63]: What does this do?

File: mountpoint-s3-fs/src/metrics_otel.rs [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+
+        Ok(Self { meter })
+    }
+
+    /// Record a counter metric in OTel format
+    pub fn record_counter(&self, key: &Key, value: u64, attributes: &[KeyValue]) {
+        let counter = self
+            .meter
+            .u64_counter(key.name().to_string())
+            .with_description("Mountpoint counter metric")
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics_otel.rs]: These are not useful descriptions.

File: mountpoint-s3-fs/src/metrics_otel.rs (hunk 8 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics_otel.rs:63]: The set_meter_provider function makes the configured meter provider the default provider that will be used throughout the application for all OpenTelemetry metrics. 

The next line obtains a meter instance from the global meter provider that was set up. The meter is used to create specific metrics (counters, gauges, histograms) and record values to them.

I'll add comments in the code to make this more clear.

File: collector-config.yaml (hunk 1 above)
Comment (15skumar) [reply collector-config.yaml]: No, it shouldn't as it was just for local testing purposes. I have deleted it now, thank you for pointing it out!

File: mountpoint-s3-fs/src/metrics.rs (hunk 2 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs]: Yes makes sense, I have added a new struct for the configuration of the otlp exporter so that fields can be specified and added in the future if needed.

File: mountpoint-s3-fs/src/metrics_otel.rs (hunk 7 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics_otel.rs]: I have now added a field in the new config struct to allow interval to be configurable. Thank you!

File: mountpoint-s3-fs/src/metrics.rs (hunk 4 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs]: The aim of this test was to send metrics to the endpoint, and I locally verified that metrics were sent by running a collector. 
Have discussed that we can have a look in the future at writing a full test to verify it within the test.

File: mountpoint-s3-fs/src/metrics.rs [hunk 10]
```diff
@@ -90,12 +93,28 @@ fn poll_process_metrics(sys: &mut System) {
 #[derive(Debug)]
 struct MetricsSink {
     metrics: DashMap<Key, Metric>,
+    otlp_exporter: Option<OtlpMetricsExporter>, // optional OTLP exporter
```
Comment (renanmagagnin) [inline mountpoint-s3-fs/src/metrics.rs]: nit: comment seems unnecessary

File: mountpoint-s3-fs/src/metrics.rs [hunk 11]
```diff
@@ -90,12 +93,28 @@ fn poll_process_metrics(sys: &mut System) {
...
 }
 
 impl MetricsSink {
-    fn new() -> Self {
+    fn new(otlp_config: Option<OtlpConfig>) -> Self {
+        // Initialise the OTLP exporter if a config is provided
+        let otlp_exporter = otlp_config
```
Comment (renanmagagnin) [inline mountpoint-s3-fs/src/metrics.rs]: We should probably handle the None otlp_config separately with it's dedicated tracing. This would provide more clarity and possibly avoid confusing with error cases

File: mountpoint-s3-fs/src/metrics.rs [hunk 12]
```diff
@@ -310,4 +360,86 @@ mod tests {
             }
         });
     }
+
+    #[test]
+    #[ignore]
```
Comment (renanmagagnin) [inline mountpoint-s3-fs/src/metrics.rs:403]: Does it make sense for us to have tests that are used for manual testing and will be ignored in the code base? Not clear what the standard for this is. If we decide to keep them, maybe we should make the purpose clear with comments

Comment (renanmagagnin) [reply mountpoint-s3-fs/src/metrics.rs:403]: Also, should we have a test for the case where the config is none?

File: mountpoint-s3-fs/src/metrics.rs (hunk 3 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs]: Since the user is explicitly requesting metrics export, it makes sense to return an error if the exporter fails to initialise for any reason. I have changed the code to do this, and have also added basic endpoint url validation.

File: mountpoint-s3-fs/src/metrics.rs (hunk 11 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs]: I have added explicit logging now for the case when no OTLP configuration is provided, thank you for the suggestion

File: mountpoint-s3-fs/src/metrics.rs (hunk 12 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs:403]: I have added comments to make the purpose of the manual tests clear, and have added a test for the case of no otlp config

File: mountpoint-s3-fs/src/metrics_otel.rs (hunk 9 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics_otel.rs]: I've removed the generic descriptions as they weren't providing specific information about the metrics and don't appear in CloudWatch. If we need descriptions in the future, we can add more specific ones that describe what each metric actually measures.

File: mountpoint-s3-fs/src/metrics.rs [hunk 13]
```diff
@@ -30,8 +33,8 @@ pub const TARGET_NAME: &str = "mountpoint_s3_fs::metrics";
 /// done with their work; metrics generated after shutting down the sink will be lost.
 ///
 /// Panics if a sink has already been installed.
-pub fn install() -> MetricsSinkHandle {
-    let sink = Arc::new(MetricsSink::new());
+pub fn install(otlp_config: Option<OtlpConfig>) -> Result<MetricsSinkHandle, String> {
```
Comment (c-hagem) [inline mountpoint-s3-fs/src/metrics.rs]: Would it make sense to have an error type here instead of returning a string? (not really sure what the convention for this would be)

**Review by c-hagem — APPROVED** [review]

File: mountpoint-s3-fs/src/metrics.rs [hunk 14]
```diff
@@ -62,9 +65,10 @@ pub fn install() -> MetricsSinkHandle {
     };
 
     let recorder = MetricsRecorder { sink };
-    metrics::set_global_recorder(recorder).unwrap();
+    metrics::set_global_recorder(recorder)
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs:68]: Would using anyhow here be better?

File: mountpoint-s3-fs/src/metrics.rs [hunk 15]
```diff
@@ -90,13 +94,45 @@ fn poll_process_metrics(sys: &mut System) {
...
 
 impl MetricsSink {
-    fn new() -> Self {
-        Self {
+    fn new(otlp_config: Option<OtlpConfig>) -> Result<Self, Box<dyn std::error::Error>> {
+        // Initialise the OTLP exporter if a config is provided
+        let otlp_exporter = if let Some(config) = otlp_config {
+            // Basic validation of the endpoint URL
+            if !config.endpoint.starts_with("http://") && !config.endpoint.starts_with("https://") {
+                return Err("Invalid OTLP endpoint configuration: endpoint must start with http:// or https://".into());
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: Same here - `anyhow` might be more readable

File: mountpoint-s3-fs/src/metrics.rs [hunk 16]
```diff
@@ -90,13 +94,45 @@ fn poll_process_metrics(sys: &mut System) {
...
+            }
+
+            match OtlpMetricsExporter::new(&config) {
+                Ok(exporter) => {
+                    tracing::info!("OpenTelemetry metrics export enabled to {}", config.endpoint);
+                    Some(exporter)
+                }
+                Err(e) => {
+                    let error_msg = format!("Failed to initialise OTLP exporter: {}", e);
+                    tracing::error!("{}", error_msg);
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs]: Any reason why this is split on two lines?

File: mountpoint-s3-fs/src/metrics.rs [hunk 17]
```diff
@@ -115,12 +151,43 @@ impl MetricsSink {
...
+                // Convert labels to OpenTelemetry KeyValue pairs
+                let attributes: Vec<KeyValue> = key
+                    .labels()
+                    .map(|label| KeyValue::new(label.key().to_string(), label.value().to_string()))
+                    .collect();
+
+                // Record the metric based on its type
+                match metric {
+                    Metric::Counter(counter) => {
+                        if let Some((value, _)) = counter.load_and_reset() {
```
Comment (muddyfish) [inline mountpoint-s3-fs/src/metrics.rs:174]: Does this interfere with how metrics are printed with the `--log-metrics` parameter?

Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs:174]: Yes it does, and I have implemented a solution for that in another branch for which I was planning to open a PR for next.

File: mountpoint-s3-fs/src/metrics.rs (hunk 14 above)
Comment (15skumar) [reply mountpoint-s3-fs/src/metrics.rs:68]: Yes using anyhow is more readable and cleaner code, i have changed the code to use anyhow for error handling now.

**Review by muddyfish — APPROVED** [review]
